lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -1,94 +1,510 @@
+
+
+class ParamType(Enum):
+    """
+    Define the possible JSON query parameter keys, and their type.
+
+    The common code can perform conversions on the required parameters with
+    known types.
+    """
+
+    DATE = ("Date", convert_date)
+    USER = ("User", convert_username)
+    JSON = ("Json", convert_json)
+    STRING = ("String", convert_string)
+
+    def __init__(self, name: AnyStr, convert: Callable[[AnyStr], Any]):
+        """
+        ParamType Enum initializer: this uses a mixed-case name string in
+        addition to the conversion method simply because with only the
+        Callable value I ran into naming issues.
+        """
+        self.friendly = name
+        self.convert = convert
+
+    def __str__(self) -> str:
+        return self.name
+
+
+class Parameter:
+    """
+    Define the attributes of a parameter using the ParamType ENUM
+
+    Note that a parameter that's "required" must also be non-empty.
+    """
+
+    def __init__(self, name: AnyStr, type: ParamType, required: bool = False):
+        """
+        __init__ Initialize a Parameter object describing a JSON parameter
+        with its type and attributes.
+
+        Args:
+            name: Parameter name
+            type: Parameter type
+            required: whether the parameter is required (default to False)
+        """
+        self.name = name
+        self.type = type
+        self.required = required
+
+    def invalid(self, json: Dict[AnyStr, Any]) -> bool:
+        """
+        ok Check whether the expression of this parameter in the JSON document
+        is invalid.
+
+        If the parameter is required and missing (or null), then it's invalid;
+        if the parameter is not required and specified as null, it's invalid.
+
+        Args:
+            json: The client JSON document being validated.
+
+        Returns:
+            True if unacceptable
+        """
+        return (self.required and self.name not in json) or (
+            self.name in json and not json[self.name]
+        )
+
+    def __str__(self) -> str:
+        return f"Parameter<{self.name}:{self.type}>"
+
+
+class Schema:
+    """
+    Define the client input schema for a server query that's based on
+    Elasticsearch.
+    """
+
+    def __init__(self, *parameters: Parameter):
+        """
+        __init__ Specify an interface schema as a list of Parameter objects.
+
+        Args:
+            parameters: a list of Parameter objects
+        """
+        self.parameters = {p.name: p for p in parameters}
+
+    def get(self, name: AnyStr) -> Parameter:
+        """
+        get Retrieve a parameter by name
+
+        Args:
+            name: Parameter name to return
+
+        Raises:
+            KeyError: The parameter wasn't found
+
+        Returns:
+            the specified Parameter value
+        """
+        if name not in self.parameters:
+            raise KeyError(f"Parameter {name} not found")
+        return self.parameters[name]
```

Comment:
This should probably be recast following the "better to apologize than to ask permission" model:
```
        try:
            return self.parameters[name]
        except KeyError:
            raise KeyError(f"Parameter {name} not found") from None
```
(Obviously, you can omit the `from None` if you think that's better.)

This avoids having to do the lookup twice in the normal/success case.

Reply:
I actually meant to delete this method, and `__iter__` since I inboarded the `validate` method and it's easier to use the `self.parameters` attribute directly rather than hiding it. I left them while I was getting it working and clearly forgot when I was done.